File: markov/markov.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

enum {
    NPREF = 2, /* number of prefix words */
    NHASH = 4093, /* size of state hash table array */
    MAXGEN = 10000 /* maximum words generated */
};

char *NONWORD = "\n"; /* cannot appear as a real word */

struct state {
    char *pref[NPREF]; /* prefix words */
    struct suffix *suf; /* list of suffixes */
    struct state *next; /* next state in hash bucket */
};

struct suffix {
    char *word; /* suffix */
    struct suffix *next; /* next in list of suffixes */
};

struct state *statetab[NHASH]; /* hash table of states */

/* hash: compute hash value for array of NPREF strings */
unsigned int hash(char *s[NPREF]) {
    unsigned int h;
    unsigned char *p;
    int i;

    h = 0;
    for (i = 0; i < NPREF; i++)
	for (p = (unsigned char *) s[i]; *p != '\0'; p++)
	    h = 31 * h + *p;
    
    return h % NHASH;
}

/*
 * lookup: search for prefix; create if requested.
 *  returns pointer if present or created; NULL if not.
 *  creation doesn't strdup so strings mustn't change later.
 */
struct state *lookup(char *prefix[NPREF], int create) {
    int i, h;
    struct state *sp;

    h = hash(prefix);

    /* look for an existing state for prefix */
    for (sp = statetab[h]; sp != NULL; sp = sp->next) {
	for (i = 0; i < NPREF; i++)
	    if (strcmp(prefix[i], sp->pref[i]) != 0)
		break;
	if (i == NPREF) /* found the right state */
	    return sp;
    }

    if (create) {
	sp = malloc(sizeof(struct state));
	for (i = 0; i < NPREF; i++)
	    sp->pref[i] = prefix[i];
	sp->suf = NULL;
	sp->next = statetab[h];
	statetab[h] = sp;
    }
    return sp;
}

/* addsuffix: add to state. suffix must not change later */
void addsuffix(struct state *sp, char *suffix) {
    struct suffix *suf;

    suf = malloc(sizeof(struct suffix));
    suf->word = suffix;
    suf->next = sp->suf;
    sp->suf = suf;
}

/* add: add word to suffix list and update prefix */
void add(char *prefix[NPREF], char *suffix) {
    struct state *sp;

    sp = lookup(prefix, 1); /* create if not found */
    addsuffix(sp, suffix);

    /* shift prefix words */
    memmove(prefix, prefix+1, (NPREF-1)*sizeof(prefix[0]));
    prefix[NPREF-1] = suffix;
}

/* build: read input and build the prefix table */
void build(char *prefix[NPREF], FILE *f) {
    char buf[100], fmt[10];

    /* create a format string to read fixed number of bytes into buf */
    sprintf(fmt, "%%%lus", sizeof(buf)-1);
    while (fscanf(f, fmt, buf) != EOF)
	add(prefix, strdup(buf));
}

/* updatepref: shift the prefix words and add a new one */
void updatepref(char *prefix[NPREF], char *w) {
    memmove(prefix, prefix+1, (NPREF-1)*sizeof(prefix[0]));
    prefix[NPREF-1] = w;
}

/* generate: produce output, one word per line */
void generate(int nwords) {
    struct state *sp;
    struct suffix *suf;
    char *prefix[NPREF], *w;
    int i, nmatch;

    for (i = 0; i < NPREF; i++) /* reset initial prefix */
	prefix[i] = NONWORD;

    for (i = 0; i < nwords; i++) {
	sp = lookup(prefix, 0);
	nmatch = 0;
	for (suf = sp->suf; suf != NULL; suf = suf->next)
	    if (rand() % ++nmatch == 0) /* prob = 1/nmatch */
		w = suf->word;
	if (strcmp(w, NONWORD) == 0)
	    break;
	printf("%s ", w);
	memmove(prefix, prefix+1, (NPREF-1)*sizeof(prefix[0]));
	prefix[NPREF-1] = w;
    }

    printf("\n");
}

int main(int argc, char *argv[]) {
    int i, nwords;
    char *prefix[NPREF];

    /* initialize random number generator */
    srand(time(NULL));

    if (argc > 1) {
	nwords = atoi(argv[1]);
    } else {
	nwords = MAXGEN;
    }
    
    for (i = 0; i < NPREF; i++) /* set up initial prefix */
	prefix[i] = NONWORD;

    build(prefix, stdin);
    add(prefix, NONWORD);
    generate(nwords);

    return 0;
}

File: markov/markov.go
package main

import (
	"fmt"
	"io"
	"math/rand"
	"os"
	"strconv"
	"time"
)

const (
	PREFIX_LEN = 2     // Markov chain prefix length
	MAX_WORDS  = 10000 // default number of words to generate
	NONWORD    = "\n"  // can't appear as a word in the input
)

var nwords = MAX_WORDS

type prefix [PREFIX_LEN]string

// emptyPrefix returns a prefix that can't appear in the input.
func emptyPrefix() prefix {
	pref := [PREFIX_LEN]string{}
	for i := 0; i < PREFIX_LEN; i++ {
		pref[i] = NONWORD
	}
	return pref
}

// newPrefix makes a new prefix from the given words.
func newPrefix(words ...string) prefix {
	pref := [PREFIX_LEN]string{}
	for i, word := range words {
		pref[i] = word
	}
	return pref
}

// addSuffix adds a word to the suffix list for pref.
func addSuffix(states map[prefix][]string, pref prefix, suf string) {
	if suffixes, ok := states[pref]; !ok {
		states[pref] = []string{suf}
	} else {
		states[pref] = append(suffixes, suf)
	}
}

// train builds a markov chain table from the text in r.
func train(r io.Reader) map[prefix][]string {
	states := make(map[prefix][]string)
	pref := emptyPrefix()
	var w string
	for n, _ := fmt.Fscan(r, &w); n > 0; n, _ = fmt.Fscan(r, &w) {
		addSuffix(states, pref, w)
		pref = newPrefix(append(pref[1:], w)...)
	}
	addSuffix(states, pref, NONWORD)
	return states
}

// generate writes nwords of text to w using the markov chain table states.
func generate(w io.Writer, states map[prefix][]string, nwords int) {
	pref := emptyPrefix()
	for i := 0; i < nwords; i++ {
		choices := states[pref]
		word := choices[rand.Intn(len(choices))]
		io.WriteString(w, fmt.Sprintf("%s ", word))
		pref = newPrefix(append(pref[1:], word)...)
	}
	io.WriteString(w, "\n")
}

func main() {
	rand.Seed(time.Now().Unix())

	if len(os.Args) > 1 {
		num, err := strconv.Atoi(os.Args[1])
		if err != nil {
			fmt.Fprintf(os.Stderr, "Invalid number of words: %v", os.Args[1])
			return
		}
		nwords = num
	}

	states := train(os.Stdin)
	generate(os.Stdout, states, nwords)
}
